### PR TITLE
small changes to server and added tests

### PR DESCRIPTION
**Pull request by BielCoelho**
- Added tests (covering 98.73% of the server)
- in PUT method the server now check if the query is empty

### CODE REVIEW

File: coverage/clover.xml [hunk 1]
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (flaviostutz) [inline coverage/clover.xml:1]: These files shouldn't be committed to repo

File: coverage/coverage-final.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
...,"line":119},"16":{"loc":{"start":{"line":123,"column":9},"end":{"line":125,"column":3}},"type":"if","locations":[{"start":{"line":123,"column":9},"end":{"line":125,"column":3}},{"start":{},"end":{}}],"line":123}},"s":{"0":1,"1":1,"2":1,"3":1,"4":6,"5":6,"6":6,"7":6,"8":6,"9":6,"10":5,"11":1,"12":1,"13":1,"14":4,"15":1,"16":1,"17":1,"18":3,"19":1,"20":1,"21":1,"22":2,"23":1,"24":1,"25":1,"26":1,"27":1,"28":1,"29":1,"30":1,"31":2,"32":1,"33":1,"34":1,"35":0,"36":1,"37":1,"38":2,"39":2,"40":27,"41":2,"42":2,"43":1,"44":14,"45":1,"46":1,"47":1,"48":1,"49":4,"50":4,"51":4,"52":4,"53":56,"54":4,"55":2,"56":2,"57":4,"58":4,"59":1,"60":14,"61":1,"62":1,"63":13,"64":1,"65":1,"66":3,"67":1,"68":2,"69":4,"70":4,"71":1,"72":2,"73":2,"74":28,"75":2,"76":28,"77":2,"78":1,"79":1,"80":1,"81":1,"82":1,"83":1,"84":1},"f":{"0":6,"1":1,"2":2,"3":27,"4":14,"5":4,"6":4,"7":56,"8":14,"9":2,"10":28,"11":28,"12":1},"b":{"0":[5,1],"1":[6,6],"2":[1,4],"3":[1,3],"4":[1,2],"5":[1,1],"6":[1,0],"7":[0,1],"8":[1,1],"9":[2,2],"10":[1,3],"11":[4,2],"12":[1,13],"13":[1,2],"14":[2,0],"15":[1,1],"16":[1,0]},"_coverageSchema":"1a1c01bbd47fc00a2c39e90264f33305004495a9","hash":"e0afdf2a24eece334cc425f4e1e7f54173cc4e35"}
```
Comment (flaviostutz) [inline coverage/coverage-final.json:1]: All coverage files should be in ignore

File: coverage/lcov-report/base.css [hunk 3]
```diff
@@ -0,0 +1,224 @@
+body, html {
```
Comment (flaviostutz) [inline coverage/lcov-report/base.css:1]: All coverage files should be in ignore

File: index.js [hunk 4]
```diff
@@ -88,28 +77,9 @@ app.get("/persons/:name", (req, res) => {
...
-    res.status(204).send(`${nameReq} person not found`);
-    return;
-  }
-
-  res.status(400).send(`<h1>400</h1><br /> <p>Bad request</p>`);
-});
-
 app.put("/persons/:name", (req, res) => {
   const { query } = req;
+  const queryExists = Object.keys(query).length > 0
```
Comment (flaviostutz) [inline index.js:82]: Why do you need query param of the identifier of the person is already part of the path?

File: index.test.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+/*
+ *   This test  needs to run fully
+ *   it creates a user to test, without the
+ *   userTest all the non-POST request will fail
+ */
+
+describe("GET to /persons", () => {
+    test("/persons returns something", async () => {
+        const response = await request(app).get("/persons")
+        expect(response.text).toEqual(expect.stringContaining(""))
```
Comment (flaviostutz) [inline index.test.js:13]: When no person is found, you shouldn't return an empty array ("[]"), not an empty string :)

File: index.test.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+});
+
+describe("GET to /persons/:user",() => {
+    test("Return user data if user exist", async () => {
+        const response = await request(app).get("/persons/userTest")
+        expect(response.statusCode).toBe(200)
+    })
+    test("Return not found if user doesn't exist", async () => {
+        const response = await request(app).get("/persons/nonExistentUser")
+        expect(response.statusCode).toBe(204)
```
Comment (flaviostutz) [inline index.test.js:83]: The correct status code for not found is "404"

File: index.test.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    })
+})
+
+describe("PUT to /persons", () => {
+  test("Updating a inexistent user is possible", async () => {
+    const response = await request(app).put(
+      "/persons/nonExistentUser?test=true"
```
Comment (flaviostutz) [inline index.test.js:90]: You should t need a query param

File: index.test.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      });
+      expect(response.statusCode).toBe(400);
+    });
+
+    test("Test if the username or height is valid", async () => {
+      const response = await request(app).post("/persons").send({
+        name: "testUser",
+        height: "180",
+      });
+      expect(response.statusCode).toBe(200);
```
Comment (flaviostutz) [inline index.test.js:38]: Normally for "created" status we use 201

File: index.test.js [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+  test("Updating a inexistent user is possible", async () => {
+    const response = await request(app).put(
+      "/persons/nonExistentUser?test=true"
+    );
+    expect(response.statusCode).toBe(404);
+  });
+
+  test("Updating a existent user is possible", async () => {
+    const response = await request(app).put("/persons/userTest?test=true");
+    expect(response.statusCode).toBe(202);
```
Comment (flaviostutz) [inline index.test.js:97]: Normally we use 200 for update.

202 is normally used when you accept the request but didn't do the effect immediately, which is not the case

File: index.test.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    expect(response.statusCode).toBe(404);
+  });
+
+  test("Updating a existent user is possible", async () => {
+    const response = await request(app).put("/persons/userTest?test=true");
+    expect(response.statusCode).toBe(202); 
+    // expect(response.body).toEqual(expect.stringContaining("successfully"))
+  });
+
+  test("Updating a existent user without query return bad query", async () => {
```
Comment (flaviostutz) [inline index.test.js:101]: I am curious why you were thinking about there query param :)

File: index.test.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+describe("DELETE to /persons", () => {
+  test("Deleting a user that doesn't exist return not found", async () => {
+    const response = await request(app).delete("/persons/nonExistentUser");
+    expect(response.statusCode).toBe(404);
+  });
+
+  test("Deleting a existent user is possible", async () => {
+    const response = await request(app).delete("/persons/testUser"); // Only works if run the POST test
+    expect(response.statusCode).toBe(202);
```
Comment (flaviostutz) [inline index.test.js:120]: Status should be 200

**Review by flaviostutz — CHANGES_REQUESTED** [review]
Very good! See that the tests are a nice way to discuss more in details the requirements of the task :)

File: coverage/clover.xml (hunk 1 above)
Comment (BielCoelho) [reply coverage/clover.xml]: forgot to add to .gitignore

File: coverage/coverage-final.json (hunk 2 above)
Comment (BielCoelho) [reply coverage/coverage-final.json]: forgot to add to .gitignore

File: coverage/lcov-report/base.css (hunk 3 above)
Comment (BielCoelho) [reply coverage/lcov-report/base.css]: forgot to add to .gitignore

File: index.js (hunk 4 above)
Comment (BielCoelho) [reply index.js:82]: in the PUT endpoint I have 3 cases
the first is accepted(the object exist + query to update exist)
the second is bad query(the object exist but there's no query)
the last is not found(object doesn't exist)

it checks if the param also come with queries

without this I couldn't test for the first case, it was ignoring the second case and going directly to the thrid case
so i made this helper const to fix this

File: index.test.js (hunk 5 above)
Comment (BielCoelho) [reply index.test.js:13]: I couldn't figure out a way to test the second case(where this path returns nothing 404), so I made this test only to see if the path returns anything, empty or a list.
I'm figuring out how to do it, so its just a placeholder for now(I should added a comment with it)

File: index.test.js (hunk 6 above)
Comment (BielCoelho) [reply index.test.js:83]: fixed

File: index.test.js (hunk 8 above)
Comment (BielCoelho) [reply index.test.js:38]: fixed

File: index.test.js (hunk 9 above)
Comment (BielCoelho) [reply index.test.js:97]: Fixed

File: index.test.js (hunk 10 above)
Comment (BielCoelho) [reply index.test.js:101]: to have a error 400 if someone tries to PUT nothing as query
without it, it would return 200

File: index.test.js (hunk 7 above)
Comment (BielCoelho) [reply index.test.js:90]: the behavior of put is always return 200 even if the object is not updated?
i did this to ensure it only return 200 if the object got update, else it return 400

File: index.test.js (hunk 5 above)
Comment (flaviostutz) [reply index.test.js:13]: Sorry. I wrote it wrong... "you should return an empty array"...

File: index.test.js (hunk 7 above)
Comment (flaviostutz) [reply index.test.js:90]: 200 only if update, 400 if a validation error occurred etc

File: index.test.js (hunk 10 above)
Comment (flaviostutz) [reply index.test.js:101]: A query param is not required

File: index.test.js (hunk 5 above)
Comment (flaviostutz) [reply index.test.js:13]: In case nothing was found

File: index.test.js (hunk 11 above)
Comment (BielCoelho) [reply index.test.js:120]: fixed

File: index.test.js (hunk 7 above)
Comment (BielCoelho) [reply index.test.js]: I finally understood what you were saying

File: index.test.js (hunk 10 above)
Comment (BielCoelho) [reply index.test.js]: I thought I couldn't send body in PUT requests

Comment (BielCoelho) [reply index.test.js]: fixed btw

**Review by flaviostutz — APPROVED** [review]
Nice!!!

**Pull request merged**